test/TEST-35-NUMAPOLICY/testsuite.sh

```diff
@@ -0,0 +1,292 @@
+#!/bin/bash
+set -ex
+set -o pipefail
+
+at_exit() {
+    if [ $? -ne 0 ]; then
+        # We're exiting with a non-zero EC, let's dump test artifacts
+        # for easier debugging
+        [ -f "$straceLog" ] && cat "$straceLog"
+        [ -f "$journalLog" ] && cat "$journalLog"
+    fi
+}
+
+trap at_exit EXIT
+
+systemd-analyze log-level debug
+systemd-analyze log-target journal
+
+# Log files
+straceLog='strace.log'
+journalLog='journal.log'
+
+# Systemd config files
+testUnit='numa-test.service'
+testUnitFile="/etc/systemd/system/$testUnit"
+testUnitNUMAConf="$testUnitFile.d/numa.conf"
+
+# Sleep constants (we should probably figure out something better but nothing comes to mind)
+journalSleep=5
+sleepAfterStart=1
+
+startStrace() {
+    coproc strace -qq -p 1 -o $straceLog -e set_mempolicy -s 1024 $1
+}
+
+stopStrace() {
+    kill -s TERM $COPROC_PID
+}
+
+startJournalctl() {
+    coproc journalctl -u init.scope -f > $journalLog
+}
+
+stopJournalctl() {
+    # Wait a few seconds until the messages get properly queued...
+    sleep $journalSleep
+    # ...and then force journald to write them to the backing storage
+    # Also, using journalctl --sync should be better than using SIGRTMIN+1, as
+    # the --sync wait until the synchronization is complete
+    echo "Force journald to write all queued messages"
+    journalctl --sync
+    kill -s TERM $COPROC_PID
+}
+
+checkNUMA() {
+    # NUMA enabled system should have at least NUMA node0
+    test -e /sys/devices/system/node/node0
+}
+
+writePID1NUMAPolicy() {
+    echo [Manager] > $confDir/numa.conf
+    echo NUMAPolicy=$1 >> $confDir/numa.conf
+    echo NUMAMask=$2>> $confDir/numa.conf
+}
+
+writeTestUnit() {
+    echo [Service] > $testUnitFile
+    echo ExecStart=/bin/sleep 3600 >> $testUnitFile
+    mkdir -p $testUnitFile.d/
+}
+
+writeTestUnitNUMAPolicy() {
+    echo [Service] > $testUnitNUMAConf
+    echo NUMAPolicy=$1 >> $testUnitNUMAConf
+    echo NUMAMask=$2>> $testUnitNUMAConf
+    systemctl daemon-reload
+}
+
+pid1ReloadWithStrace() {
+    startStrace
+    systemctl daemon-reload
+    stopStrace
+}
+
+pid1ReloadWithJournal() {
+    startJournalctl
+    systemctl daemon-reload
+    stopJournalctl
+}
+
+pid1StartUnitWithStrace() {
+    startStrace '-f'
+    systemctl start $1
+    sleep $sleepAfterStart
+    stopStrace
+}
+
+pid1StartUnitWithJournal() {
+    startJournalctl
+    systemctl start $1
+    sleep $sleepAfterStart
+    stopJournalctl
+}
+
+pid1StopUnit() {
+    systemctl stop $1
+}
+
+systemctlCheckNUMAProperties() {
+    local LOGFILE="$(mktemp)"
+    systemctl show -p NUMAPolicy $1 > "$LOGFILE"
+    grep "NUMAPolicy=$2" "$LOGFILE"
+
+    > "$LOGFILE"
+
+    if [ -n $3 ]; then
+        systemctl show -p NUMAMask $1 > "$LOGFILE"
+        grep "NUMAMask=$3" "$LOGFILE"
+    fi
+}
+
+checkNUMA
```

Comment:
Should we just exit here if the check fails? See e.g. `TEST-19-DELEGATE/testsuite.sh`.

Reply:
I think that's exactly what @mrc0mmand meant by "I guess it would make sense to skip the test when NUMA is not detected, instead of failing".

Reply:
On the other hand, the test could probably check whether whatever happens on the systemd side when NUMA isn't supported makes any sense. Currently by just looking at the test, it's not clear whether `systemd` ignores the settings or fails to start services.

Reply:
Maybe a little bit OT (and I'm probably overthinking it), but wouldn't it make sense to incorporate a special exit code just for skipped tests, like we already do with unit tests (EC 77)? Right now, this (and, for example, the mentioned `TEST-19-DELEGATE`) appears to be passing even on systems without appropriate support, which might be unfortunate in some situations.

Reply:
I don't think you're overthinking it. After all it's common practice to signal that some tests are skipped. On the other hand, in practice 77 is indistinguishable from 0 in the sense that when tests don't fail loudly usually nobody looks at them.